hadoop-hdds/docs/content/design/storage-class.md

```diff
@@ -19,10 +20,408 @@ author: Marton Ele
+    * For normal `RATIS` replication only the `ratis_quorum`, and the `required_closed_replicas` can be configured and only two tuples are valid `(ONE,ONE)`, `(THREE,THREE)`
+    * `STANDALONE` replication is deprecated but we need to support it until full removal (without parameters)
+ * `StorageClass` is a POJO with the selection of `ReplicationMechanism` and the replication configuration.
+ * We need a `StorageClassRegistry` instance which is available both from client side and server side. Today we don't need to make the configurable from external files: both the existing replication mechanisms can be configured from the code. 
+ * `StorageClassRegistry` will contain the definition of the three initial storage-classes: STANDARD (RATIS/THREE), REDUCED (RATIS/ONE) and LEGACY (STANDALONE)
+ * We need a *legacy mapping table* which maps the `factor/type` from any older requests to `storageClass`! This can help to keep the backward compatibility.
+  
+**Protocol changes**: 
+
+  * `storageClass` field should be added to all the protocols where we use `factor/type` today. Old `factor/type` fields will be kept for compatibility reasons. 
+  * Storage-class will be represented as a custom String in the RPC protocol (in storage-sensitive places it can be stored as an integer with an additional mapping).
+  * There are two protocols which should be changed:
+   * *Key creation* and *Key information* related structures in `OmClientProtocol` should be updated (`KeyInfo`, `MultipartUploadInfo`, `KeyArgs`,... )
+   *  SCMAdmin protocol should be improved to include `storage-class` in `ContainerRequestProto`.
+
+**Ozone Client changes**: Ozone client classes (`OzoneBucket`, `OzoneKey`,...) should be modified to propagate the `storageClass` everywhere instead of `factor/type`. To make it backward compatible we can create legacy mapping table should be used.
+
+**Ozone CLI changes** `ozone sh` CLI will be changed to support `--storage-class` instead of `--type` and `--factor`
+
+**HCFS changes**: `ofs://` and `o3fs://` doesn't require any modification
+
+**Ozone Manager changes**: OM needs the following changes:
+
+  * `storageClass` should be stored for the keys instead of `factor/type`. THe *legacy mapping table* will be used to generate storageClass for old, persisted keys which has `factory/type` in the RocksDB but no storageClass. It can be added to the appropriate codec.
+  * `storageClass` should be propagated when SCM is called from OM for new container allocations.
+  * If `storageClass` is not defined for a key, the default should be used from the `bucket`/`volume`. If they are not specified the cluster-wide default will be used.
+
+**Storage Container Manager** changes:
+
+  * `ContainerInfo` will include the `storageClass` as a property
+  * Internal allocation request will contain the `storageClass` (old requests can be handled with the *legacy mapping table*)
```

Comment:
How will SCM make decisions about allocating pipelines of different kinds?

Reply:
We don't need to modify any logic in SCM. We already have allocation logic for Ratis/Standalone pipelines. When a new request comes in, we should choose from the pre-allocated pipelines OR create a new one, based on replication factor.

This logic is exactly the same with storage-class. Based on the proposed new storage-class field of the `AllocateScmBlock` can identify the required replication scheme (eg. Ratis) and parameters (eg. THREE) and choose the pipeline exactly the same way s today.

Obviously, EC requires new pipeline management, which is not covered in this document (as suggested).

But it shows the benefit of the storage-class: the list of configured storage-class provides a pre-defined list of the required pipelines. For example if two storage classes are defined asEC 6:3 and EC 3:2, SCM will have the information about the required pipelines to pre-allocate.